src/python/pants/core/util_rules/archive.py

```diff
@@ -28,38 +36,83 @@ class ExtractedDigest:
     digest: Digest
 
 
-def get_extraction_cmd(archive_path: str, output_dir: str) -> Optional[Tuple[str, ...]]:
-    """Returns a shell command to run to extract the archive to the output directory."""
-    # Note that we assume that mkdir, unzip, and tar exist in the executing environment.
-    if archive_path.endswith(".zip"):
-        return ("unzip", "-q", archive_path, "-d", output_dir)
-    elif archive_path.endswith((".tar", ".tar.gz", ".tgz", ".tar.bz2", ".tbz2", ".tar.xz", ".txz")):
-        return ("mkdir", output_dir, "&&", "tar", "xf", archive_path, "-C", output_dir)
-    return None
+# TODO: Should this be configurable?
+SEARCH_PATHS = ("/usr/bin", "/bin", "/usr/local/bin")
+
+
+class UnzipBinary(BinaryPath):
+    def extract_argv(self, *, archive_path: str, output_dir: str) -> Tuple[str, ...]:
+        # Note that the `output_dir` does not need to already exist. The caller should also
+        # validate that it's a valid `.zip` file.
+        return (self.path, "-q", archive_path, "-d", output_dir)
+
+
+@rule(level=LogLevel.DEBUG)
+async def find_unzip() -> UnzipBinary:
+    request = BinaryPathRequest(
+        binary_name="unzip", search_path=SEARCH_PATHS, test=BinaryPathTest(args=["-v"])
+    )
+    paths = await Get(BinaryPaths, BinaryPathRequest, request)
+    first_path = paths.first_path
+    if not first_path:
+        raise BinaryNotFoundError(request, rationale="download the tools Pants needs to run")
+    return UnzipBinary(first_path.path, first_path.fingerprint)
+
+
+class TarBinary(BinaryPath):
+    def extract_argv(self, *, archive_path: str, output_dir: str) -> Tuple[str, ...]:
+        # Note that the `output_dir` must already exist. The caller should also
+        # validate that it's a valid `.tar` file.
+        return (self.path, "xf", archive_path, "-C", output_dir)
```

Comment:
Wouldn't a `z` argument also be necessary if the file is compressed?

Reply:
I honestly don't know. I copied this from Benjy's code. Should I add it?

Reply:
Does it work in tests without the `z`?

Reply:
I didn't need it to get these tests passing:

https://github.com/pantsbuild/pants/blob/2fd32f02bd2cb26a1f3fffffee1a2880a0d5631d/src/python/pants/core/util_rules/archive_test.py#L43-L59

Is this what you mean, or is there a different type of compression you have in mind?

Reply:
No it was more about whether `tar` detected compressed archives automatically. Sounds like it does?

Reply:
Yeah, I think so.